### DISCUSSION TITLE
Milvus 1.x - Too many open segments in the collection.

### DISCUSSION THREAD

**Discussion by durai89** [milvus-io/milvus#18598] [Q&A and General discussion] [answered]
We have a real time ingestion pipeline that pushes vectors to Milvus. We are noticing too many open segments in the collection, and because they haven't met the `index_file_size` threshold none of the segments have been indexed affecting query latencies.

I am trying to understand
- What leads to too many open segments? Why a new segment gets created though `index_file_size` threshold is not met on the already open segments?
- How can we fix them? We understand compact operation can merge smaller growing segments in Milvus 2.x, but do we have any equivalent in Milvus 1.x?

Any ideas to close/merge these small open segments and trigger index build is much appreciated! Thanks

**Comment by yhmo** [comment, marked as answer]
In v1.x
- When you call insert(), the inserted data firstly is put into "insert buffer"
- When you call flush(), the data in "insert buffer" will be flushed into hard disk. For each collection, each flush() action will generate a small segment
- If you didn't call flush() manually, the milvus will automatically call flush() with interval(by default 1 second)
- A background thread of milvus will do compaction if it found small segments, merge small segments into large segment, until the large segment size exceed index_file_size
- to create index, typically we have two ways:
   1. insert all vectors ---> call create_index(), in this way, when you call create_index(), this api will be blocked util all segments are indexed successfullhy, all segments which rows larger than 1024 will be indexed. 
   2. call create_index() ---> streaming insert vectors, in this way, create_index() do nothing and return, you specify an index type, milvus automatically build index for segment which size larger than index_file_size
- once a segment is to be indexed, it will not take part in compaction even its size is smaller than index_file_size.

**Comment by yhmo** [comment]
Do some test to observe segment count:
Firstly disable auto-flush behavior, in the server_config.yml, set auto_flush_interval to be 0.
```
storage:
  auto_flush_interval: 0
```

Test 1:
1. create a collection, with index_file_size=100MB
2. call insert()+flush() 10 times, each insert request input 10MB data, totally 100MB data
3. call create_index()

Each flush() generate a new segment, so the step 2 will generate 10 segments, each segment size is 10MB.
The creae_index() will internally do a compaction before creating index, the 10 segments is merged into a large segment which size is 100MB, then milvus build index for this large segment. Finally there is only one segment.

Test 2:
1. create collection, with index_file_size=100MB
2. call insert()+flush()+create_index() 10 times, each insert request input 10MB data, totally 100MB data

The step 2 call create_index() after each insert(), each flush() generate a segment(size=10MB), then create_index() will build index for this segment, since this segment is indexed, will not take part in compaction. So, finally there are 10 segments in the storage.
